# Week-2

In [8]:
import pandas as pd
# 1. Read merged listing dataset
import pandas as pd
df1 = pd.read_csv("CRMLSSold_merge.csv", low_memory=False)
# Append Check 
# =========================
print("=== APPEND CHECK ===")
print("Row count of combined/appended dataset:", df1.shape[0])

=== APPEND CHECK ===
Row count of combined/appended dataset: 591733


In [3]:
# 2. Unique property types
print("\n=== UNIQUE PROPERTY TYPES ===")
property_type_counts = df1["PropertyType"].value_counts(dropna=False)
print(property_type_counts)

property_type_table = property_type_counts.reset_index()
property_type_table.columns = ["PropertyType", "Count"]


=== UNIQUE PROPERTY TYPES ===
PropertyType
Residential            397603
ResidentialLease       135617
Land                    19345
ManufacturedInPark      16082
ResidentialIncome       15865
CommercialSale           3714
CommercialLease          3104
BusinessOpportunity       403
Name: count, dtype: int64


In [4]:
df = pd.read_csv("CRMLSSold_residential.csv")
print("\n=== PropertyType Check ===")
display(df["PropertyType"].value_counts(dropna=False))

/var/folders/d_/gll3jlg1445_2nm3xgq_xvdc0000gn/T/ipykernel_29367/3488275968.py:1: DtypeWarning: Columns (0: BuyerAgentAOR, 1: ListAgentAOR, 2: ListAgentEmail, 3: FireplaceYN, 4: OriginatingSystemName, 5: OriginatingSystemSubName, 6: BuyerAgencyCompensationType, 7: latfilled, 8: lonfilled) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("CRMLSSold_residential.csv")



=== PropertyType Check ===


PropertyType
Residential    397603
Name: count, dtype: int64

In [5]:
# 4. Missing value summary
missing_summary = pd.DataFrame({
    "column": df.columns,
    "null_count": df.isnull().sum().values,
    "null_pct": (df.isnull().mean().values * 100).round(2)
})

missing_summary["flag_above_90pct_null"] = missing_summary["null_pct"] > 90
missing_summary = missing_summary.sort_values(by="null_count", ascending=False)

In [6]:
print("\n=== MISSING VALUE SUMMARY ===")
display(missing_summary)

print("\n=== COLUMNS ABOVE 90% NULL ===")
high_null_cols = missing_summary[missing_summary["flag_above_90pct_null"]]
display(high_null_cols)


=== MISSING VALUE SUMMARY ===


,column,null_count,null_pct,flag_above_90pct_null
62,CoveredSpaces,397603,100.0,True
77,MiddleOrJuniorSchoolDistrict,397603,100.0,True
32,AboveGradeFinishedArea,397603,100.0,True
30,FireplacesTotal,397603,100.0,True
51,TaxYear,397603,100.0,True
...,...,...,...,...
20,DaysOnMarket,0,0.0,False
21,ListOfficeName,0,0.0,False
48,ListingId,0,0.0,False
33,ListingKeyNumeric,0,0.0,False



=== COLUMNS ABOVE 90% NULL ===


,column,null_count,null_pct,flag_above_90pct_null
62,CoveredSpaces,397603,100.00,True
77,MiddleOrJuniorSchoolDistrict,397603,100.00,True
32,AboveGradeFinishedArea,397603,100.00,True
30,FireplacesTotal,397603,100.00,True
51,TaxYear,397603,100.00,True
55,ElementarySchoolDistrict,397603,100.00,True
60,BusinessType,397603,100.00,True
35,TaxAnnualAmount,397603,100.00,True
4,WaterfrontYN,397355,99.94,True
59,BelowGradeFinishedArea,395313,99.42,True


In [7]:
# 5. Numeric distribution summary
target_numeric_cols = ["ListPrice", "LivingArea", "DaysOnMarket"]

summary_rows = []

for col in target_numeric_cols:
    if col in df.columns:
        s = pd.to_numeric(df[col], errors="coerce")
        summary_rows.append({
            "column": col,
            "count_non_null": s.notna().sum(),
            "min": s.min(),
            "p5": s.quantile(0.05),
            "p25": s.quantile(0.25),
            "median": s.median(),
            "mean": s.mean(),
            "p75": s.quantile(0.75),
            "p95": s.quantile(0.95),
            "max": s.max()
        })
    else:
        summary_rows.append({
            "column": col,
            "count_non_null": None,
            "min": None,
            "p5": None,
            "p25": None,
            "median": None,
            "mean": None,
            "p75": None,
            "p95": None,
            "max": None
        })

numeric_summary = pd.DataFrame(summary_rows)

print("\n=== NUMERIC DISTRIBUTION SUMMARY ===")
display(numeric_summary)


=== NUMERIC DISTRIBUTION SUMMARY ===


,column,count_non_null,min,p5,p25,median,mean,p75,p95,max
0,ListPrice,397603,525.0,346500.0,575000.0,815000.0,1.138630e+06,1295000.0,2850000.0,137500000.0
1,LivingArea,397374,0.0,839.0,1247.0,1641.0,1.904351e+03,2217.0,3558.0,17021321.0
2,DaysOnMarket,397603,-288.0,1.0,8.0,19.0,3.733679e+01,48.0,131.0,12430.0


The numerical summary indicates that all three variables—listing price, living area, and days on market—exhibit right-skewness, as their mean values ​​exceed their respective medians.

Regarding the listing price, the median stands at $815,000, while the mean is notably higher at approximately $1.14 million; this suggests that a small number of high-priced properties have inflated the average, pointing to the presence of extreme outliers.

In terms of living area, the median is 1,641 square feet, with a mean of approximately 1,904 square feet, indicating a moderately right-skewed distribution. However, this observation may be influenced by outliers or potential data quality issues.

As for "days on market," the median is 19 days, while the mean is approximately 37 days; this suggests that while most properties sell relatively quickly, a subset remains on the market for extended periods. Furthermore, the presence of a negative minimum value (–288) is illogical, indicating a data error that requires correction during the data cleaning process.

# summary

Data overiew
The combined dataset total： 591733 rows
After filtering (PropertyType = Residential): 397603 rows
Date: January 2024 to March 2026

●	What is the Residential vs. other property type share? 

In [15]:
prop_counts = df["PropertyType"].value_counts(dropna=False)
total = len(df)

print("\n=== PropertyType Distribution ===")
print(prop_counts)
prop_pct = (prop_counts / total * 100).round(2)

print("\n=== PropertyType Share (%) ===")
print(prop_pct)



=== PropertyType Distribution ===
PropertyType
Residential    397603
Name: count, dtype: int64

=== PropertyType Share (%) ===
PropertyType
Residential    100.0
Name: count, dtype: float64


Residential properties account for 100% of the dataset, while other property types account for 0%. This indicates that the dataset has already been fully filtered to include only residential properties.

●	What are the median and average close prices? 


In [17]:

mean_price = df["ClosePrice"].mean()
median_price = df["ClosePrice"].median()

print("\n=== ClosePrice Summary ===")
print(f"Average (Mean): {round(mean_price, 2)}")
print(f"Median: {round(median_price, 2)}")
print("\n=== ClosePrice Distribution ===")
print(df["ClosePrice"].describe())


=== ClosePrice Summary ===
Average (Mean): 1185616.36
Median: 820000.0

=== ClosePrice Distribution ===
count    3.976010e+05
mean     1.185616e+06
std      5.922380e+06
min      0.000000e+00
25%      5.750000e+05
50%      8.200000e+05
75%      1.300000e+06
max      9.895000e+08
Name: ClosePrice, dtype: float64


The average (mean) close price is approximately $1,185,616, while the median close price is $820,000.

●	What does the Days on Market distribution look like? 

The Days on Market distribution appears to be right-skewed. Most properties are sold relatively quickly, as shown by the lower median, but there are some listings that stay on the market much longer. These longer listings create a long tail and increase the average.

●	What percentage of homes sold above vs. below list price? 

In [19]:
above = (df["ClosePrice"] > df["ListPrice"]).sum()
below = (df["ClosePrice"] < df["ListPrice"]).sum()
equal = (df["ClosePrice"] == df["ListPrice"]).sum()

total = len(df)

above_pct = round(above / total * 100, 2)
below_pct = round(below / total * 100, 2)
equal_pct = round(equal / total * 100, 2)

print("\n=== Sold Price vs List Price ===")
print(f"Above List Price: {above_pct}%")
print(f"Below List Price: {below_pct}%")
print(f"Equal to List Price: {equal_pct}%")


=== Sold Price vs List Price ===
Above List Price: 40.11%
Below List Price: 42.54%
Equal to List Price: 17.34%


Approximately 40.11% of homes sold above the list price, while 42.54% sold below. The remaining 17.34% of homes sold at exactly the list price.

Are there any apparent date consistency issues (e.g., close date before listing date)? ●	

In [20]:

df["PurchaseContractDate"] = pd.to_datetime(df["PurchaseContractDate"], errors="coerce")


close_before_list = (df["CloseDate"] < df["ListingContractDate"]).sum()
purchase_before_list = (df["PurchaseContractDate"] < df["ListingContractDate"]).sum()
close_before_purchase = (df["CloseDate"] < df["PurchaseContractDate"]).sum()

total = len(df)

print("\n=== Date Consistency Check ===")
print(f"Close < Listing: {close_before_list} ({round(close_before_list/total*100,4)}%)")
print(f"Purchase < Listing: {purchase_before_list} ({round(purchase_before_list/total*100,4)}%)")
print(f"Close < Purchase: {close_before_purchase} ({round(close_before_purchase/total*100,4)}%)")


=== Date Consistency Check ===
Close < Listing: 58 (0.0146%)
Purchase < Listing: 261 (0.0656%)
Close < Purchase: 240 (0.0604%)


A very small number of records show date inconsistencies, such as close or purchase dates occurring before the listing date. These issues affect less than 0.1% of the dataset and are likely due to data entry errors, so they should be addressed during data cleaning.

●	Which counties have the highest median prices? 


In [28]:

county_median = (
    df.groupby("CountyOrParish")["ClosePrice"]
    .median()
    .sort_values(ascending=False))

print("\n=== Median Close Price by County ===")
print(county_median.head(10))


=== Median Close Price by County ===
CountyOrParish
Del Norte        2485000.0
Other County     2462500.0
San Mateo        1700000.0
Santa Clara      1600000.0
Santa Cruz       1200000.0
San Francisco    1180000.0
Orange           1175000.0
Marin            1170000.0
Alameda          1135000.0
Alpine           1100000.0
Name: ClosePrice, dtype: float64
